libraries/c_sdk/standard/ble/src/iot_ble_mqtt_transport.c

```diff
@@ -0,0 +1,1081 @@
+    }
+    else
+    {
+        pBuffer->pBuffer[ 0 ] = packetType;
+        pBuffer->pBuffer[ 1 ] = 2;
+        pBuffer->pBuffer[ 2 ] = ( uint8_t ) ( packetId >> 8 );
+        pBuffer->pBuffer[ 3 ] = ( uint8_t ) ( packetId & 0x00ffU );
+    }
+
+    return status;
+}
+
+
+
+static MQTTStatus_t transportSerializeSuback( MQTTFixedBuffer_t * pBuffer,
+                                              uint8_t packetType,
+                                              uint16_t packetId )
+{
+    MQTTStatus_t status = MQTTSuccess;
+
+    if( pBuffer == NULL )
+    {
+        LogError( ( "Provided buffer is NULL." ) );
+        status = MQTTBadParameter;
+    }
+    /* The buffer must be able to fit 4 bytes for the packet. */
+    else if( pBuffer->size < SIZE_OF_SUB_ACK )
+    {
+        LogError( ( "Insufficient memory for packet." ) );
+        status = MQTTNoMemory;
+    }
+    else if( packetId == 0U )
+    {
+        LogError( ( "Packet ID cannot be 0." ) );
+        status = MQTTBadParameter;
+    }
+    else
+    {
+        pBuffer->pBuffer[ 0 ] = packetType;
+        pBuffer->pBuffer[ 1 ] = 3;
+        pBuffer->pBuffer[ 2 ] = ( uint8_t ) ( packetId >> 8 );
+        pBuffer->pBuffer[ 3 ] = ( uint8_t ) ( packetId & 0x00ffU );
+        pBuffer->pBuffer[ 4 ] = 1;
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static MQTTStatus_t handleOutgoingConnect( const void * buf,
+                                           MQTTFixedBuffer_t * bufToSend,
+                                           size_t * pSize )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    MQTTConnectInfo_t connectConfig;
+
+    status = parseConnect( &connectConfig, buf, pSize );
+
+    if( status == MQTTSuccess )
+    {
+        status = IotBleMqtt_SerializeConnect( &connectConfig, &bufToSend->pBuffer, pSize );
+    }
+
+    return status;
+}
+
+static MQTTStatus_t handleOutgoingPublish( const void * buf,
+                                           MQTTFixedBuffer_t * bufToSend,
+                                           size_t * pSize )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    uint16_t packetIdentifier = 0;
+    MQTTPublishInfo_t publishConfig;
+
+    status = parsePublish( &publishConfig, buf, pSize, &packetIdentifier );
+
+    if( status == MQTTSuccess )
+    {
+        status = IotBleMqtt_SerializePublish( &publishConfig, &bufToSend->pBuffer, pSize, &packetIdentifier, NULL );
+    }
+
+    return status;
+}
+
+
+static MQTTStatus_t handleOutgoingPuback( const void * buf,
+                                          MQTTFixedBuffer_t * bufToSend,
+                                          size_t * pSize )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    uint16_t packetIdentifier = 0;
+
+    status = parsePubAck( buf, &packetIdentifier );
+
+    if( status == MQTTSuccess )
+    {
+        /* Pub Ack packet sizes are always 4 bytes (control + remaining length + 2 identifier) */
+        *pSize = SIZE_OF_SIMPLE_ACK;
```

Comment:
IotBleMqtt_Serialize functions will calculate the required size and allocate the buffer for you. so you don't need to pass the size.